sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -23,6 +23,15 @@ import org.apache.spark.sql.catalyst.trees
 import org.apache.spark.sql.catalyst.trees.TreeNode
 import org.apache.spark.sql.types._
 
+
+/**
+ * For Catalyst to work correctly, concrete implementations of [[Expression]]s must be case classes
+ * whose constructor arguments are all Expressions types. In addition, if we want to support more
```

Comment:
So that when the expression is used with `FunctionRegistry.expression`, the constructor (actually an `apply` method of the companion object) can be found via reflection.